api/explorer/api/ingest.py

```diff
@@ -101,89 +101,100 @@ def _edge_from_build_input_type(build_input_type: model.BuildInputType) -> graph
         )
 
 
-def split_nix_graph(
-    nix_graph: model.NixGraph,
+class _IngestionContext:
+    """Class holding data that is passed along successive calls to
+    split_nix_package"""
+
+    def __init__(self) -> None:
+        """Map from package output path to graph Package object.
+        Continuously grows while ingesting"""
+        self.packages: dict[str, graph.Package] = {}
+
+        """Edges that could not be added yet, because the package
+           that is the destination of the edge has not been processed yet.
+           For example if package A at "/A" depends on "/B" and package "A" is
+           processed first, this dict will be {"/B": [(A, edge)]} until B is processed.
+           When package B is processed, "A -edge-> B" will be returned in the
+           ingestion methods below, and the dict is shrunk.
+
+           This dict has the destination of the edge as key, to allow for fast
+           lookup (as opposed to using a list of triplets)."""
+        # TODO, this is a multimap, is there a Python class for that?
+        self.pending_edges: dict[str, list[tuple[graph.Package, graph.Edge]]] = {}
```

Comment:
Can we use [dataclasses](https://docs.python.org/3/library/dataclasses.html) instead?
Maybe even immutable (`frozen=True`)?

Reply:
Added an additional commit to use `dataclasses` (not sure the added boilerplate is worth it to be honest, but why not) :+1: